klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,921 @@
+            exit(1)
+
+
+def get_switches(yaml_file, rule_name):
+    """Parse yaml file and extract switches data
+    Parameters
+    ----------
+    yaml_file : str
+            yaml config file path given py the user.
+    Returns
+    -------
+    yaml_dic : dictionary
+            dictionary containing switches data.
+    """
+
+    # load yaml config data
+    with open(yaml_file, 'r') as stream:
+        try:
+            yaml_dic = yaml.safe_load(stream)
+        except yaml.YAMLError as exc:
+            print(exc)
+
+    switches = list()
+    for param, value in yaml_dic[rule_name].items():
+        switch = f"{param}={value}"
+        switches.append(switch)
+
+    return switches
+
+
+def parse_results_db(results_database):
+    """
+    This function will parse Klayout database for analysis.
+
+    Parameters
+    ----------
+    results_database : string or Path object
+        Path string to the results file
+
+    Returns
+    -------
+    set
+        A set that contains all rules in the database with violations
+    """
+
+    mytree = ET.parse(results_database)
+    myroot = mytree.getroot()
+
+    # Initial values for counter
+    rule_counts = defaultdict(int)
+
+    for z in myroot[7]:
+        rule_name = f"{z[1].text}".replace("'", "")
+        rule_counts[rule_name] += 1
+
+    return rule_counts
+
+
+def run_test_case(
+    drc_dir,
+    layout_path,
+    run_dir,
+    table_name,
+):
+    """
+    This function run a single test case using the correct DRC file.
+
+    Parameters
+    ----------
+    drc_dir : string or Path
+        Path to the location where all runsets exist.
+    layout_path : stirng or Path object
+        Path string to the layout of the test pattern we want to test.
+    run_dir : stirng or Path object
+        Path to the location where is the regression run is done.
+    table_name : string
+        Table name that we are running on.
+
+    Returns
+    -------
+    dict
+        A dict with all rule counts
+    """
+
+    # Initial value for counters
+    rule_counts = defaultdict(int)
+
+    # Get switches used for each run
+    sw_file = os.path.join(Path(layout_path.parent).absolute(), f"{table_name}.{SUPPORTED_SW_EXT}")
+
+    if os.path.exists(sw_file):
+        switches = " ".join(get_switches(sw_file, table_name))
+    else:
+        switches = "--variant=C"  # default switch
+
+    # Adding switches for specific runsets
+    if "antenna" in str(layout_path):
+        switches += " --antenna_only"
+    elif "density" in str(layout_path):
+        switches += " --density_only"
+
+    # Creating run folder structure
+    pattern_clean = ".".join(os.path.basename(layout_path).split(".")[:-1])
+    output_loc = f"{run_dir}/{table_name}"
+    pattern_log = f"{output_loc}/{pattern_clean}_drc.log"
+
+    # command to run drc
+    call_str = f"python3 {drc_dir}/run_drc.py --path={layout_path} {switches} --table={table_name} --run_dir={output_loc} --run_mode=flat --thr=1  > {pattern_log} 2>&1"
```

Comment:
any reason we don't `import run_drc` here and run it as a function?

Reply:
The intention of the regression to mimic the user usage. The user will not import run_drc rather they will just call the run_drc from commandline.

Reply:
one of the issue with this is that you can less structured error reporting and add additional hurdle to troubleshoot (you wouldn't be able to easily step thru this with a debugger for example)

Reply:
@proppy I understand your concerns. But I believe the use case of debugger here is not going to be very important. Keep in mind, that even the `run_drc.py` is just a script that calls klayout at the end. From my perspective you want your tests to reflect the actual use case that the user would do.

But if you want me to change that I don't mind. Please let me know what you want me to do?

Reply:
maybe simply calling `main` could be a good middleground?

Reply:
Will be done as described in the issue https://github.com/google/globalfoundries-pdk-libs-gf180mcu_fd_pv/issues/6

Reply:
Here I'm suggesting to import and call the main function rather shelling out (which seems unrelated to #6).

Reply:
Added #10 to cover this. Could we close this PR on the current status?